core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/CompositeRuinMoveSelector.java

```diff
@@ -0,0 +1,266 @@
+            for (Object entity : entities) {
+                for (GenuineVariableDescriptor<Solution_> variableDescriptor : variableDescriptorList) {
+                    if (variableDescriptor.isGenuineListVariable()) {
+                        ListVariableDescriptor<Solution_> listVariableDescriptor =
+                                (ListVariableDescriptor<Solution_>) variableDescriptor;
+                        List<Object> listVariable = listVariableDescriptor.getListVariable(entity);
+                        for (int sourceIndex = listVariable.size() - 1; sourceIndex >= 0; --sourceIndex) {
+                            moves.add(new ListUnassignMove<>(listVariableDescriptor, entity, sourceIndex));
+                        }
+                    } else if (!variableDescriptor.isChained()) {
+                        moves.add(new ChangeMove<>(variableDescriptor, entity, null));
+                    }
+                }
+
+                for (ShadowVariableDescriptor<Solution_> shadowVariableDescriptor : shadowVariableDescriptors) {
+                    if (shadowVariableDescriptor instanceof InverseRelationShadowVariableDescriptor) {
+                        Object sourceEntity = shadowVariableDescriptor.getValue(entity);
+                        if (sourceEntity != null) {
+                            List<VariableDescriptor<Solution_>> sourceVariableDescriptors =
+                                    shadowVariableDescriptor.getSourceVariableDescriptorList();
+                            for (VariableDescriptor<Solution_> sourceVariableDescriptor : sourceVariableDescriptors) {
+                                if (sourceVariableDescriptor.isGenuineListVariable()) {
+                                    ListVariableDescriptor<Solution_> listVariableDescriptor =
+                                            (ListVariableDescriptor<Solution_>) sourceVariableDescriptor;
+                                    List<Object> listVariable = listVariableDescriptor.getListVariable(sourceEntity);
+                                    int sourceIndex = listVariable.indexOf(entity);
+                                    listUnassignMoves
+                                            .add(new ListUnassignMove<>(listVariableDescriptor, sourceEntity, sourceIndex));
+                                } else {
+                                    moves.add(new ChangeMove<>((GenuineVariableDescriptor<Solution_>) sourceVariableDescriptor,
+                                            sourceEntity, null));
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            Comparator<ListUnassignMove<Solution_>> listUnassignMoveComparator =
+                    Comparator.comparing(ListUnassignMove::getSourceIndex);
+            listUnassignMoves.sort(listUnassignMoveComparator.reversed());
+            moves.addAll(listUnassignMoves);
```

Comment:
Please add a comment describing why it's necessary to reverse the order of unassign moves. It's a ruin move - why do we care in which order is the solution ruined?